### ISSUE TITLE
[BUG BOUNTY] $500 Reward to improve the performance of Meteor on Windows

### ISSUE THREAD

**Issue by wreiske** [meteor/meteor#12935]
# **Please clearly read the success criteria before working on this bounty.**

### $500 Bug Bounty: Meteor.js Performance Improvements on Windows

**Objective:** To address and improve the performance issues encountered in Meteor when running on Windows, especially concerning `meteor update`, package installations, builds, rebuilds, and other development processes.

**Bounty Reward:** $500

### Current State
Videos of current state can be found here: https://drive.google.com/drive/folders/1_OeBoJxg1X_sbVfSl9sVitUPMdBppD0a?usp=sharing and the timings below

| Command      | Windows 11 | Debian 12 | Video
| ----------- | ----------- | ----------- | ----------- |
| npm install -g meteor      | 10 min 3 sec       | 26.664 sec | [Installing Meteor First Time.mkv](https://drive.google.com/file/d/1Tf42ZkNFjdIAUUZRJI_-gtuFJd67VYr8/view?usp=drive_link) | 
| meteor create testapp --blaze   | 44.29 sec        |  12.96 sec | [Creating a new meteor project.mkv](https://drive.google.com/file/d/1gcw_o2sn5wR55vY3HuVmXNWUoOEULarM/view?usp=drive_link)
| meteor  | 45.87 sec        |  24.19 sec | [Starting a fresh meteor project for the first time.mkv](https://drive.google.com/file/d/1x_QD2W05Hg8eW0vwHjxM7wtg-PvPlxZo/view?usp=drive_link)
| meteor add ostrio:flow-router-extra  | 14.27 sec        |  9.61 sec | [Adding a meteor package.mkv](https://drive.google.com/file/d/1eW_xww5-lGj3ZWwNJm_IxlZBMhm2qw2L/view?usp=drive_link)
| meteor update --release 3.0-alpha.19  | **Gave up (see below)**        |  7 min 2 sec | [Updating meteor release.mkv](https://drive.google.com/file/d/1W9SasS_o1hcJLpB54RaZDbZhgQr9sxb2/view?usp=drive_link)

The update for meteor didn't finish after an hour 37 minutes on windows. The 7 minutes for debian included the time it took me to switch node versions, install make, and a few other dependecies, and start up the meteor project on the new release. 

#### Success Criteria

1. **Root Cause Analysis:** 
   - Identify the specific reasons contributing to the slowness and intermittent failures of the `meteor update` on Windows.
   - Provide a detailed analysis of how these issues differ from the performance on Linux and MacOS.

2. **Performance Improvement:**
   - Implement solutions that significantly improve the performance of `meteor update` and package installations on Windows.
   - Ensure that these improvements do not negatively impact the performance on other operating systems (Linux and MacOS).

3. **Testing and Validation:**
   - Demonstrate through testing that the improvements are effective and consistent across various Windows environments.
   - Include benchmarks or metrics showing the performance improvement compared to the previous version.

4. **Documentation and Reporting:**
   - Provide comprehensive documentation of the changes made, including technical details and reasons for specific approaches.
   - Outline steps for reproducing the improvements and any necessary environment setups or configurations.

5. **Maintainability and Compatibility:**
   - Ensure that the improvements are maintainable and do not introduce new bugs or vulnerabilities.
   - Verify compatibility with the latest version of Meteor.js (3.0) and widely-used packages.

#### Submission Guidelines

- Submit a pull request to the Meteor.js GitHub repository with your changes. These changes should be compatible with the latest target version of meteor 3.0
- Include detailed commit messages and comments explaining your code changes.
- Attach a report summarizing your approach, findings, and testing methodology.

#### Evaluation Criteria

- Effectiveness of the solution in improving performance on Windows.
- Quality of analysis and documentation.
- Adherence to Meteor.js coding standards and practices.
- Compatibility and non-disruptiveness to existing Meteor.js functionality.

### Related Issues
https://github.com/meteor/meteor/discussions/12648#discussioncomment-7903122
https://github.com/meteor/meteor/issues/12633
https://github.com/meteor/meteor/issues/12459
https://github.com/meteor/meteor/issues/11308
https://github.com/meteor/meteor/issues/11149
https://github.com/meteor/meteor/issues/10881
https://github.com/meteor/meteor/issues/10601
https://github.com/meteor/meteor/issues/10322
https://github.com/meteor/meteor/issues/9592

### Sponsor Notice

This bounty is being paid for by my employer, Medical Informatics Engineering, in an effort to help give back to the meteor community. We're also [hiring](https://www.mieweb.com/careers).

**Comment by mrkirthi-24** [comment]
what's the deadline to submit the solution??

**Comment by wreiske** [comment]
> what's the deadline to submit the solution??

I missed adding that to the original description. The bounty will be live for 6 months until Sunday, June 23, 2024. Thank you.

**Comment by mrkirthi-24** [comment]
> I missed adding that to the original description. The bounty will be live for 6 months until Sunday, June 23, 2024. Thank you.

Thank you @wreiske.

**Comment by ayewo** [comment]
@wreiske 
Just to be sure, this issue is focused on doing a RCA of the slowness and intermittent failure of the `meteor update` command {_1. research_}, fixing the performance issue(s) {_2. performance fix_}, show that performance is improved {_3. benchmarks / regression tests_} and explaining what was done {_4. report_}, correct?

Explicitly doing the same activities {_1, 2, 3, 4_} for the other CLI commands used for illustrating the performance issue above i.e.
- `npm i -g meteor`, 
- `meteor create …`, 
- `meteor` and 
- `meteor add …` is not in scope, correct?

I imagine the performance fix for `meteor update…`
will likely resolve most of the performance issues observed in the other commands mentioned and if not, the report’s findings will generally point you in the direction you should be looking to fix them. Is my understanding correct?

**Comment by wreiske** [comment]
> @wreiske Just to be sure, this issue is focused on doing a RCA of the slowness and intermittent failure of the `meteor update` command {_1. research_}, fixing the performance issue(s) {_2. performance fix_}, show that performance is improved {_3. benchmarks / regression tests_} and explaining what was done {_4. report_}, correct?
> 
> Explicitly doing the same activities {_1, 2, 3, 4_} for the other CLI commands used for illustrating the performance issue above i.e.
> 
> * `npm i -g meteor`,
> * `meteor create …`,
> * `meteor` and
> * `meteor add …` is not in scope, correct?
> 
> I imagine the performance fix for `meteor update…` will likely resolve most of the performance issues observed in the other commands mentioned and if not, the report’s findings will generally point you in the direction you should be looking to fix them. Is my understanding correct?

The success criteria mention two things specifically: Implement solutions that 1. significantly improve the performance and stability of meteor update and 2. package installations on Windows.

My biggest frustration always comes when doing a `meteor update` and a `meteor update --all-packages`. Running a `meteor add package-name` may share a lot of the same underlying code as the `meteor update` when it runs package updates, so yes, it would be safe to imagine RCA of `meteor update` performance will likely also help understand other commands. Running `meteor` for the first time on a project that you don't have packages for will also be much slower on Windows because it has to go and fetch all of the dependencies. It's my assumption that would also share some of the underlying code for packages that should be covered in the RCA of the performance issues and whatever fixes are required to improve the performance.

**Comment by ayewo** [comment]
> Yes to everything except it should also include the performance fix of package installs. I wouldn't consider each command needing the full RCA process.

Thanks for the confirmation. 

(In a lot of cases, merely switching from `npm` to a performance oriented alternative like `yarn` or `pnpm`, which are known to aggressively de-duplicate dependencies during installation, can shave considerable time off package installs. It’s fine if switching package managers is not an option.)

**Comment by harryadel** [comment]
Glad to see folks giving back to the Meteor community. Kudos to you @wreiske and your employer.

**Comment by harryadel** [comment]
This needs to be promoted a hella lot more. 

@Grubba27 @acamikuro 
official channels: dev.to, X/twitter & blog

@wreiske 
A forum post, and nobody should've the honors to do it but you.

**Comment by wreiske** [comment]
@harryadel , thanks for the suggestion. I've posted to the forums here: https://forums.meteor.com/t/bug-bounty-500-reward-to-improve-the-performance-of-meteor-on-windows/61236

**Comment by rizzling** [comment]
still open?  or has @wreike or someone else took the bounty already ?

**Comment by harryadel** [comment]
@rizzling It doesn't seem to have even been [attempted before](https://console.algora.io/org/meteor/bounties?status=open) @rizzling It'd be awesome if you can give it a go!

**Comment by rizzling** [comment]
> @rizzling It doesn't seem to have even been [attempted before](https://console.algora.io/org/meteor/bounties?status=open) @rizzling It'd be awesome if you can give it a go!

okay cool, I think I'm gonna give it a try. I will let y'all know if I do so with /attempt ;)

**Comment by ayewo** [comment]
> 
> I found the code that might be slowing it down. Running the commands as Administrator causes it to use 7zip for decompression. Also, adding the folder the npm package is in to Windows Defender exclusion, to exclude it from antivirus checks, may speed it up too.
> 
> I first attempted to install as non-admin, and it was taking forever. I then reran the command as Administrator, and it only took a minute to install. I ran the "meteor update" command and it only took two minutes to update.
> 
> Adding code to the package, and asking for admin rights, to get the ability to use symlinks, would fix this. Also add code to add folders to Windows Defender exclusion list.
> 
> https://github.com/meteor/meteor/blob/02aa69c0c2d2c3ea805ba8bcf0dbaa565ca995b3/npm-packages/meteor-installer/install.js#L263
> 
> ![Screenshot 2024-05-08 194012](https://private-user-images.githubusercontent.com/1155115/328892376-93676a78-ed1a-4f34-8260-d6f935b60df8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUxNzYzMjIsIm5iZiI6MTcxNTE3NjAyMiwicGF0aCI6Ii8xMTU1MTE1LzMyODg5MjM3Ni05MzY3NmE3OC1lZDFhLTRmMzQtODI2MC1kNmY5MzViNjBkZjgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MDhUMTM0NzAyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTJiMWIwNWYxMzM0NWUwMzQ4OWM1ZjQ2MTg1ZDcyMTdhNzE3NjJhYmRmYzJjMTZlNDhjODM2NTdiMmYzNWE3MiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.66vCyg-ez30IMfG4Nvp6KUGD_9Lyge-OBO_qKJFdEK8)
> 
> ![Screenshot 2024-05-08 193942](https://private-user-images.githubusercontent.com/1155115/328892441-ea9a01ec-240f-4bd2-a714-5d347299b8e8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUxNzYzMjIsIm5iZiI6MTcxNTE3NjAyMiwicGF0aCI6Ii8xMTU1MTE1LzMyODg5MjQ0MS1lYTlhMDFlYy0yNDBmLTRiZDItYTcxNC01ZDM0NzI5OWI4ZTgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MDhUMTM0NzAyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjllOGFlN2RmODQxNzhlYzZjNDUzNWIxYWNmM2FmZTU5ZmIyMWFiZGMwYWU5ZDFlZWFlNmI2OGVmYjc1M2UxYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.ILkqE12X6MsgF-LG0Vk2gIRMdbd4wjSJgfLBCoxKZhc)
> 
> ![Screenshot 2024-05-08 194120](https://private-user-images.githubusercontent.com/1155115/328892540-fbd48379-a5b7-421f-ab36-4ba1bd3794e4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUxNzYzMjIsIm5iZiI6MTcxNTE3NjAyMiwicGF0aCI6Ii8xMTU1MTE1LzMyODg5MjU0MC1mYmQ0ODM3OS1hNWI3LTQyMWYtYWIzNi00YmExYmQzNzk0ZTQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MDhUMTM0NzAyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjhkNmVkMDVlNThhZTRkNTRhZDFmODE2Y2ZlMjMyNTI0MjM5NzA1MmZmZDAyMGYwNjY0NTVjZWVjZTMzYThiOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.MJJ8ZQD3QnqUjqnBTdj2kPaFR1Dso85eTg6LGMYkXg4)
> 
> ![Screenshot 2024-05-08 194024](https://private-user-images.githubusercontent.com/1155115/328892583-83a9cd6f-1920-45d4-ab33-070eeadfeaa0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUxNzYzMjIsIm5iZiI6MTcxNTE3NjAyMiwicGF0aCI6Ii8xMTU1MTE1LzMyODg5MjU4My04M2E5Y2Q2Zi0xOTIwLTQ1ZDQtYWIzMy0wNzBlZWFkZmVhYTAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MDhUMTM0NzAyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDAxZmUzYzQ3NmQ4YmZlZTg0MGY2MjJlNzJkYWQ2NGM0NDM2MGE5YjczMTc4YjIyZTQ1ZjM0ODA2NDA5ZDRmYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Rq1SuckeCwEr74dlgdM8ZT_qfdxOSlGjkbF7f5uQs_I)
> 
> Options

Hey @leonewton253
I did some preliminary work on it as well but got side-tracked by other stuff until I saw the email notification from your attempt. 

1. Yes, adding folder/process exclusions for meteor in Windows Defender was one of the first things I tried that lead to a significant speed up. I confirmed with Will over email that he didn’t have any exclusions in place when he recorded the execution times above.

2. There is an additional speed up for the `npm install` command if you switch out the Node.js-based `tar` implementation for the`bsdtar` executable.

Now that you have kicked the cobwebs off this issue, not sure whether to publish the report I was working on as-is, or try to finish it before making it public.

**Comment by ayewo** [comment]
Hi @wreiske,

I've spent significant amounts of time researching the root cause of the slowness on Window and it boils down to 2 factors:
1. heavy interference from Windows Defender which scans files as they are extracted by different binaries  i.e. `npm`/`meteor`/`7zip` etc;
2. the design of `meteor` which [assumes](https://github.com/meteor/meteor/blob/fc1fcfd999008c89166ee4b0745ae2c0878a293e/tools/PERFORMANCE.md#hardware-and-os) a UNIX-style filesystem for everything. This leads to extra work (read: additional I/O calls) on Windows in order to preserve this UNIX-style filesystem abstraction for `meteor` which is why [Windows is slower](https://github.com/meteor/meteor/tree/fc1fcfd999008c89166ee4b0745ae2c0878a293e/tools/fs#files-vs-fs-and-filespath-vs-path).

The only way to isolate the effects of factor 1. above is to completely disable Windows Defender so it is easy to see how `meteor` commands on Windows truly compare to macOS and Linux and I did that.

Using the 5 commands you used in your initial investigation, I wrote some benchmarking code (in Terraform) that logs the execution times on Linux (I used Ubuntu 22.04 LTS) and on Windows (I used Windows 2022 with Windows Defender completely disabled) and these are the results: 

|   | Command                               | Linux    | Windows (Defender❌)  |
|---|---------------------------------------|---------------------|---------------------|
|1  | `npm install -g meteor`               | 19.137s  | 108.624s | 
|2  | `meteor create testapp --blaze`       | 48.594s | 70.036s  | 
|3  | `meteor`                              | 21.077s | 25.177s  | 
|4  | `meteor add ostrio:flow-router-extra` | 8.476s | 10.913s  | 
|5  | `meteor update --release 3.0-alpha.19`| 2min 11.139s (131.139s) | 11min 30.851s (690.851s)  |

I used a [c5a.2xlarge](https://instances.vantage.sh/aws/ec2/c5a.2xlarge) instance type for the benchmarks and as you can see, the time differences aren't as wide as [yours](https://github.com/meteor/meteor/issues/12935#issue-2055383824), though of course Windows is still slower overall.

**Comment by ayewo** [comment]
@wreiske 
Just a quick note that I've opened a [PR](https://github.com/meteor/meteor/pull/13203) within the Meteor repo as stated in the issue. 

You might want to start with the [report](https://github.com/ayewo/meteor-windows-performance/blob/main/README.md) which summarizes my approach, findings, and testing methodology.

**Comment by wreiske** [comment]
@ayewo , amazing work! I will take a look this weekend and get back to you. 🙏

**Comment by leonewton253** [comment]
> Hi @wreiske,
> 
> I've spent significant amounts of time researching the root cause of the slowness on Window and it boils down to 2 factors:
> 
>     1. heavy interference from Windows Defender which scans files as they are extracted by different binaries  i.e. `npm`/`meteor`/`7zip` etc;
> 
>     2. the design of `meteor` which [assumes](https://github.com/meteor/meteor/blob/fc1fcfd999008c89166ee4b0745ae2c0878a293e/tools/PERFORMANCE.md#hardware-and-os) a UNIX-style filesystem for everything. This leads to extra work (read: additional I/O calls) on Windows in order to preserve this UNIX-style filesystem abstraction for `meteor` which is why [Windows is slower](https://github.com/meteor/meteor/tree/fc1fcfd999008c89166ee4b0745ae2c0878a293e/tools/fs#files-vs-fs-and-filespath-vs-path).
> 
> 
> The only way to isolate the effects of factor 1. above is to completely disable Windows Defender so it is easy to see how `meteor` commands on Windows truly compare to macOS and Linux and I did that.
> 
> Using the 5 commands you used in your initial investigation, I wrote some benchmarking code (in Terraform) that logs the execution times on Linux (I used Ubuntu 22.04 LTS) and on Windows (I used Windows 2022 with Windows Defender completely disabled) and these are the results:
> 	Command 	Linux 	Windows (Defender❌)
> 1 	`npm install -g meteor` 	19.137s 	108.624s
> 2 	`meteor create testapp --blaze` 	48.594s 	70.036s
> 3 	`meteor` 	21.077s 	25.177s
> 4 	`meteor add ostrio:flow-router-extra` 	8.476s 	10.913s
> 5 	`meteor update --release 3.0-alpha.19` 	2min 11.139s (131.139s) 	11min 30.851s (690.851s)
> 
> I used a [c5a.2xlarge](https://instances.vantage.sh/aws/ec2/c5a.2xlarge) instance type for the benchmarks and as you can see, the time differences aren't as wide as [yours](https://github.com/meteor/meteor/issues/12935#issue-2055383824), though of course Windows is still slower overall.

I was able to get the "meteor update --release 3.0-alpha.19" command close to 2 mins on my windows machine using Cygwin terminal. Maybe @wreiske can check that out, in conjunction with turning defender off, or adding the folder exclusion.
[Cygwin](https://www.cygwin.com/)

**Comment by wreiske** [comment]
Something else I've noticed recently is that if updates seem hung on Windows, you can Control+C the process, re-run it, and it will complete faster. 

For example, I am currently upgrading to 3.0-rc.8 and it was stuck on standard-minifier-css for a long time, over 10 minutes. I Control+C, up-arrow, enter, and it completed in under a minute.

**Comment by ayewo** [comment]
Hey @wreiske 

Sorry that I haven't been able to respond. 
I've been backed up a bit but should have some availability early next week, God willing.

**Comment by wreiske** [comment]
I'd love to get any of this progress merged into meteor and pay out the bounty if meteor maintainers have reviewed the fixes and agree there are no breaking changes.

**Comment by ayewo** [comment]
Same here. 

I've addressed all of the feedback raised in the [PR](https://github.com/meteor/meteor/pull/13203) but I guessing merging it is not a priority for the team right now.

Other than that, I just need to find some quiet time to look into the issue you mentioned [above](https://github.com/meteor/meteor/issues/12935#issuecomment-2227506697).

**Comment by ayewo** [comment]
@wreiske If you are open to paying out part of the bounty while we wait on the Meteor team to merge, whenever that may be, that’s also fine too.

The Algora command is:
/tip $<amount> @ayewo

**Comment by denihs** [comment]
Hi, between today and tomorrow, I'll take a look at everything here.

**Comment by ayewo** [comment]
Hi @wreiske 

With respect to your expectations in the PR [discussion](https://github.com/meteor/meteor/pull/13203#issuecomment-2307735612):

> For the sake of the bounty, I was really hoping for the `create` and `update` and `add` and other tools to have better performance. At the very least, some sort of automation to pop an administrator prompt up to "whitelist" the directories required to get better performance, or if administrator prompt is needed for proper symbolic links, that should be implemented.

If you don't mind, I'd like to continue the discussion here now that the PR has been merged.

1. As regards whitelisting, I've added significantly more [process and folder exclusions](https://github.com/ayewo/meteor-windows-performance/blob/28f671d5667709884ee6202f111ed671e28d0f3a/windows/02_process-and-folder-exclusions.ps1#L26-L85). The whitelist is now comprehensive so it should mitigate the effects of any slow down from Defender. 

2. > or if administrator prompt is needed for proper symbolic links, that should be implemented.

With respect to this, simply enabling Developer Mode on Windows 11 will actually suppress the administrator prompt whenever an app needs to create proper symbolic links[^1]. I confirmed it worked [some months ago](https://github.com/ayewo/test-symlinks-on-windows10) so no special carve out is required in the Power Shell script.

Please let me know what you think.


[^1]: On Windows, Meteor [appears](https://github.com/search?q=repo%3Ameteor%2Fmeteor%20junction&type=code) to mostly use directory junctions. Permission wise, directory junctions are more robust compared to symlinks because the user doesn't need administrator privileges to create them, though of course but they are less flexible than symbolic links because they can't point to files (they can only point to directories) or link to a folder on another volume (i.e. a directory junction created on drive `C:\` can't refer to `D:\npm-cache` because the folder is on a different drive).

**Comment by ayewo** [comment]
I've added a convenience [script](https://github.com/ayewo/meteor-windows-performance/blob/main/defender-exclusions.ps1) you can use to automate the whitelisting during development.

1. If you run it without arguments, it will just add predefined exclusions:
`.\defender-exclusions.ps1`

2. If you'd like to whitelist a specific folder, you can simply do:`.\defender-exclusions.ps1 Add C:\path\to\folder`.
As an example, assuming the following folder layout:
```bash
cd C:\User\Administrator\projects
meteor create testapp-two --blaze
```
Running `.\defender-exclusions.ps1 Add C:\User\Administrator\projects` will add that folder to the exclusions along with the predefined ones.

**Comment by ayewo** [comment]
Hey @wreiske 
what else do I need to do to push this over the finish line :) ?
